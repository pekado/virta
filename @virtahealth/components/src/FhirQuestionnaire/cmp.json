{
  "status": "draft",
  "resourceType": "Questionnaire",
  "meta": {
    "profile": [
      "http://hl7.org/fhir/us/sdc/StructureDefinition/sdc-questionnaire|2.0"
    ],
    "tag": [
      {
        "code": "lformsVersion: 24.0.0"
      }
    ]
  },
  "item": [
    {
      "type": "group",
      "code": [
        {
          "code": "24323-8",
          "display": "Comprehensive metabolic 2000 panel - Serum or Plasma",
          "system": "http://loinc.org"
        }
      ],
      "required": false,
      "linkId": "/24323-8",
      "text": "Comprehensive metabolic 2000 panel - Serum or Plasma",
      "item": [
        {
          "type": "decimal",
          "code": [
            {
              "code": "2345-7",
              "display": "Glucose SerPl-mCnc",
              "system": "http://loinc.org"
            }
          ],
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
              "valueCoding": {
                "display": "mg/dL"
              }
            }
          ],
          "required": false,
          "linkId": "/24323-8/2345-7",
          "text": "Glucose SerPl-mCnc"
        },
        {
          "type": "group",
          "code": [
            {
              "code": "24326-1",
              "display": "Lytes 1998 Pnl SerPl",
              "system": "http://loinc.org"
            }
          ],
          "required": false,
          "linkId": "/24323-8/24326-1",
          "text": "Lytes 1998 Pnl SerPl",
          "item": [
            {
              "type": "decimal",
              "code": [
                {
                  "code": "2951-2",
                  "display": "Sodium SerPl-sCnc",
                  "system": "http://loinc.org"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
                  "valueCoding": {
                    "display": "mmol/L"
                  }
                }
              ],
              "required": false,
              "linkId": "/24323-8/24326-1/2951-2",
              "text": "Sodium SerPl-sCnc"
            },
            {
              "type": "decimal",
              "code": [
                {
                  "code": "2823-3",
                  "display": "Potassium SerPl-sCnc",
                  "system": "http://loinc.org"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
                  "valueCoding": {
                    "display": "mmol/L"
                  }
                }
              ],
              "required": false,
              "linkId": "/24323-8/24326-1/2823-3",
              "text": "Potassium SerPl-sCnc"
            },
            {
              "type": "decimal",
              "code": [
                {
                  "code": "2075-0",
                  "display": "Chloride SerPl-sCnc",
                  "system": "http://loinc.org"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
                  "valueCoding": {
                    "display": "mmol/L"
                  }
                }
              ],
              "required": false,
              "linkId": "/24323-8/24326-1/2075-0",
              "text": "Chloride SerPl-sCnc"
            },
            {
              "type": "decimal",
              "code": [
                {
                  "code": "1963-8",
                  "display": "HCO3 SerPl-sCnc",
                  "system": "http://loinc.org"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
                  "valueCoding": {
                    "display": "mmol/L"
                  }
                }
              ],
              "required": false,
              "linkId": "/24323-8/24326-1/1963-8",
              "text": "HCO3 SerPl-sCnc"
            },
            {
              "type": "decimal",
              "code": [
                {
                  "code": "2028-9",
                  "display": "CO2 SerPl-sCnc",
                  "system": "http://loinc.org"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
                  "valueCoding": {
                    "display": "mmol/L"
                  }
                }
              ],
              "required": false,
              "linkId": "/24323-8/24326-1/2028-9",
              "text": "CO2 SerPl-sCnc"
            },
            {
              "type": "decimal",
              "code": [
                {
                  "code": "33037-3",
                  "display": "Anion Gap SerPl-sCnc",
                  "system": "http://loinc.org"
                }
              ],
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
                  "valueCoding": {
                    "display": "mmol/L"
                  }
                }
              ],
              "required": false,
              "linkId": "/24323-8/24326-1/33037-3",
              "text": "Anion Gap SerPl-sCnc"
            },
            {
              "text": "The components of this panel were defined by HCFA (now CMS)",
              "type": "display",
              "linkId": "/24323-8/24326-1-help",
              "extension": [
                {
                  "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
                  "valueCodeableConcept": {
                    "text": "Help-Button",
                    "coding": [
                      {
                        "code": "help",
                        "display": "Help-Button",
                        "system": "http://hl7.org/fhir/questionnaire-item-control"
                      }
                    ]
                  }
                }
              ]
            }
          ]
        },
        {
          "text": "The components of this panel were defined by HCFA (now CMS)",
          "type": "display",
          "linkId": "/24323-8-help",
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-itemControl",
              "valueCodeableConcept": {
                "text": "Help-Button",
                "coding": [
                  {
                    "code": "help",
                    "display": "Help-Button",
                    "system": "http://hl7.org/fhir/questionnaire-item-control"
                  }
                ]
              }
            }
          ]
        }
      ]
    }
  ]
}
