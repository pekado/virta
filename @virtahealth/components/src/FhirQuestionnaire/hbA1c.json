{
  "status": "draft",
  "resourceType": "Questionnaire",
  "meta": {
    "profile": [
      "http://hl7.org/fhir/us/sdc/StructureDefinition/sdc-questionnaire|2.0"
    ],
    "tag": [
      {
        "code": "lformsVersion: 24.0.0"
      }
    ]
  },
  "item": [
    {
      "type": "group",
      "code": [],
      "required": false,
      "linkId": "/4548-4",
      "text": "Hgb A1c MFr Bld",
      "item": [
        {
          "type": "decimal",
          "code": [
            {
              "code": "4548-4",
              "display": "Hgb A1c MFr Bld",
              "system": "http://loinc.org"
            }
          ],
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
              "valueCoding": {
                "display": "%"
              }
            }
          ],
          "required": false,
          "linkId": "/4548-4",
          "text": "Hgb A1c MFr Bld"
        }
      ]
    }
  ]
}
