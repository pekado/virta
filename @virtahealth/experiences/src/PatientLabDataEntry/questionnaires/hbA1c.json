{
  "status": "draft",
  "resourceType": "Questionnaire",
  "meta": {
    "profile": [
      "http://hl7.org/fhir/us/sdc/StructureDefinition/sdc-questionnaire|2.0"
    ],
    "tag": [
      {
        "code": "lformsVersion: 24.0.0"
      }
    ]
  },
  "item": [
    {
      "type": "group",
      "code": [],
      "required": false,
      "linkId": "/4548-4",
      "text": "Hemoglobin A1c",
      "item": [
        {
          "type": "decimal",
          "code": [
            {
              "code": "4548-4",
              "display": "HbA1c",
              "system": "http://loinc.org"
            }
          ],
          "extension": [
            {
              "url": "http://hl7.org/fhir/StructureDefinition/questionnaire-unit",
              "valueCoding": {
                "display": "%"
              }
            },
            {
              "url": "http://hl7.org/fhir/StructureDefinition/minValue",
              "valueDecimal": 4.0
            },
            {
              "url": "http://hl7.org/fhir/StructureDefinition/maxValue",
              "valueDecimal": 15.5
            }
          ],
          "required": false,
          "linkId": "/4548-4",
          "text": "HbA1c"
        }
      ]
    }
  ]
}
