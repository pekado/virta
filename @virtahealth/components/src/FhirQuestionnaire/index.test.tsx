import * as React from "react";
import { Text } from "react-native";
import { fireEvent } from "@testing-library/react-native";
import { createComponentWithVirtaContext } from "../test";
import * as hbA1CQuestionnaire from "./hbA1c.json";
import * as partialCmpQuestionnaire from "./cmp.json";
import { FhirQuestionnaire } from "./index";

describe("FHIR Questionnaire component", () => {
  it("renders children", () => {
    const { getByText } = createComponentWithVirtaContext(
      <FhirQuestionnaire
        questionnaire={{ status: "draft" }}
        setQuestionnaireResponse={() => undefined}
      >
        <Text>hello</Text>
      </FhirQuestionnaire>
    );
    expect(getByText("hello")).toBeTruthy();
  });

  it("renders questionnaire", () => {
    const fhirQuestionnaire = createComponentWithVirtaContext(
      <FhirQuestionnaire
        questionnaire={{
          status: "draft",
          item: [
            {
              code: [{ code: "1234-5", display: "sample question" }],
              linkId: "test-linkId",
              type: "question",
            },
            {
              code: [{ code: "5432-1", display: "sample group of questions" }],
              linkId: "test-linkId2",
              type: "group",
              item: [
                {
                  code: [{ code: "2345-6", display: "sample group question" }],
                  linkId: "test-string-linkId",
                  type: "string",
                },
                {
                  code: [
                    {
                      code: "2345-6",
                      display: "another sample group question",
                    },
                  ],
                  linkId: "test-decimal-linkId",
                  type: "decimal",
                },
              ],
            },
          ],
        }}
        setQuestionnaireResponse={() => undefined}
      />
    );
    expect(fhirQuestionnaire).toMatchSnapshot();
  });

  it("updates responses appropriately", () => {
    const testSetQuestionnaireResponse = jest.fn();
    const hbA1cQuestionnaireView = createComponentWithVirtaContext(
      <FhirQuestionnaire
        questionnaire={hbA1CQuestionnaire}
        setQuestionnaireResponse={testSetQuestionnaireResponse}
        questionnaireResponse={{
          item: [
            {
              item: [],
              linkId: "/4548-4",
            },
          ],
          status: "in-progress",
        }}
      />
    );
    const hbA1cInput = hbA1cQuestionnaireView.getByTestId("/4548-4");
    fireEvent.changeText(hbA1cInput, "6.5");
    expect(testSetQuestionnaireResponse).toHaveBeenCalledWith({
      status: "in-progress",
      item: [
        {
          linkId: "/4548-4",
          item: [
            {
              linkId: "/4548-4",
              answer: [{ valueDecimal: "6.5" }],
            },
          ],
        },
      ],
    });
  });

  describe("with nested group items", () => {
    it("updates response appropriately", () => {
      const testSetQuestionnaireResponse = jest.fn();
      const cmpQuestionnaireView = createComponentWithVirtaContext(
        <FhirQuestionnaire
          questionnaire={partialCmpQuestionnaire}
          setQuestionnaireResponse={testSetQuestionnaireResponse}
          questionnaireResponse={{
            item: [
              {
                item: [],
                linkId: "/24323-8",
              },
            ],
            status: "in-progress",
          }}
        />
      );
      const cmpInput = cmpQuestionnaireView.getByTestId(
        "/24323-8/24326-1/2951-2"
      );
      fireEvent.changeText(cmpInput, "6.5");
      expect(testSetQuestionnaireResponse).toHaveBeenCalledWith({
        status: "in-progress",
        item: [
          {
            linkId: "/24323-8",
            item: [
              {
                linkId: "/24323-8/24326-1/2951-2",
                answer: [{ valueDecimal: "6.5" }],
              },
            ],
          },
        ],
      });
    });
  });
});
