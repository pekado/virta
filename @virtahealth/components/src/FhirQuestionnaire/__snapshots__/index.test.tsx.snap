// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FHIR Questionnaire component renders questionnaire 1`] = `
<View>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "flex-start",
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "marginRight": 32,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "display": "flex",
              "flexBasis": "auto",
              "flexDirection": "column",
              "flexGrow": 1,
              "flexShrink": 0,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "marginRight": 0,
                "marginTop": 0,
                "minWidth": 205,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "rgb(36, 38, 43)",
                  "fontFamily": "Whitney-Book",
                  "fontSize": 12,
                  "fontWeight": "500",
                  "letterSpacing": 0,
                  "lineHeight": 12,
                },
              ]
            }
          >
            sample question
          </Text>
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "flex-start",
                },
                Object {
                  "display": "flex",
                  "marginTop": 10,
                },
              ]
            }
          >
            <View
              editable={true}
              isFocused={false}
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  },
                  Object {
                    "alignItems": "center",
                    "backgroundColor": "rgb(255,255,255)",
                    "borderColor": "rgb(214, 217, 221)",
                    "borderRadius": 6,
                    "borderStyle": "solid",
                    "borderWidth": 2,
                    "display": "flex",
                    "flexBasis": "auto",
                    "flexGrow": 1,
                    "flexShrink": 0,
                  },
                ]
              }
            >
              <TextInput
                editable={true}
                isFocused={false}
                onBlur={[Function]}
                onChangeText={[Function]}
                onFocus={[Function]}
                style={
                  Array [
                    Object {
                      "backgroundColor": "transparent",
                      "color": "rgb(0,0,0)",
                      "flexBasis": 0,
                      "flexGrow": 1,
                      "flexShrink": 1,
                      "fontFamily": "Whitney-Book",
                      "fontSize": 20,
                      "minHeight": 40,
                      "paddingBottom": 8,
                      "paddingLeft": 12,
                      "paddingRight": 12,
                      "paddingTop": 8,
                    },
                  ]
                }
                testID="test-linkId"
                value=""
              />
            </View>
          </View>
        </View>
      </View>
      <View
        style={
          Array [
            Object {
              "height": 36,
            },
          ]
        }
      />
    </View>
  </View>
  <View>
    <Text
      style={
        Array [
          Object {
            "color": "rgb(36, 38, 43)",
            "fontFamily": "Whitney-Medium",
            "fontSize": 20,
            "fontWeight": "500",
            "letterSpacing": 0,
            "lineHeight": 20,
          },
          Object {
            "borderBottomWidth": 1,
            "borderColor": "rgb(214, 217, 221)",
            "fontWeight": "600",
            "lineHeight": 28,
            "marginBottom": 32,
            "paddingBottom": 12,
            "paddingTop": 12,
          },
        ]
      }
    >
      sample group of questions
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "flexWrap": "wrap",
            "justifyContent": "flex-start",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "marginRight": 32,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "display": "flex",
                "flexBasis": "auto",
                "flexDirection": "column",
                "flexGrow": 1,
                "flexShrink": 0,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "marginRight": 0,
                  "marginTop": 0,
                  "minWidth": 205,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "rgb(36, 38, 43)",
                    "fontFamily": "Whitney-Book",
                    "fontSize": 12,
                    "fontWeight": "500",
                    "letterSpacing": 0,
                    "lineHeight": 12,
                  },
                ]
              }
            >
              sample group question
            </Text>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  },
                  Object {
                    "display": "flex",
                    "marginTop": 10,
                  },
                ]
              }
            >
              <View
                editable={true}
                isFocused={false}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                    },
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "rgb(255,255,255)",
                      "borderColor": "rgb(214, 217, 221)",
                      "borderRadius": 6,
                      "borderStyle": "solid",
                      "borderWidth": 2,
                      "display": "flex",
                      "flexBasis": "auto",
                      "flexGrow": 1,
                      "flexShrink": 0,
                    },
                  ]
                }
              >
                <TextInput
                  editable={true}
                  isFocused={false}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "color": "rgb(0,0,0)",
                        "flexBasis": 0,
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "fontFamily": "Whitney-Book",
                        "fontSize": 20,
                        "minHeight": 40,
                        "paddingBottom": 8,
                        "paddingLeft": 12,
                        "paddingRight": 12,
                        "paddingTop": 8,
                      },
                    ]
                  }
                  testID="test-string-linkId"
                  value=""
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "height": 36,
              },
            ]
          }
        />
      </View>
      <View
        style={
          Array [
            Object {
              "marginRight": 32,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "display": "flex",
                "flexBasis": "auto",
                "flexDirection": "column",
                "flexGrow": 1,
                "flexShrink": 0,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "marginRight": 0,
                  "marginTop": 0,
                  "minWidth": 205,
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "rgb(36, 38, 43)",
                    "fontFamily": "Whitney-Book",
                    "fontSize": 12,
                    "fontWeight": "500",
                    "letterSpacing": 0,
                    "lineHeight": 12,
                  },
                ]
              }
            >
              another sample group question
            </Text>
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexDirection": "row",
                    "justifyContent": "flex-start",
                  },
                  Object {
                    "display": "flex",
                    "marginTop": 10,
                  },
                ]
              }
            >
              <View
                editable={true}
                isFocused={false}
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "justifyContent": "flex-start",
                    },
                    Object {
                      "alignItems": "center",
                      "backgroundColor": "rgb(255,255,255)",
                      "borderColor": "rgb(214, 217, 221)",
                      "borderRadius": 6,
                      "borderStyle": "solid",
                      "borderWidth": 2,
                      "display": "flex",
                      "flexBasis": "auto",
                      "flexGrow": 1,
                      "flexShrink": 0,
                    },
                  ]
                }
              >
                <TextInput
                  editable={true}
                  isFocused={false}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  style={
                    Array [
                      Object {
                        "backgroundColor": "transparent",
                        "color": "rgb(0,0,0)",
                        "flexBasis": 0,
                        "flexGrow": 1,
                        "flexShrink": 1,
                        "fontFamily": "Whitney-Book",
                        "fontSize": 20,
                        "minHeight": 40,
                        "paddingBottom": 8,
                        "paddingLeft": 12,
                        "paddingRight": 12,
                        "paddingTop": 8,
                      },
                    ]
                  }
                  testID="test-decimal-linkId"
                  value=""
                />
              </View>
            </View>
          </View>
        </View>
        <View
          style={
            Array [
              Object {
                "height": 36,
              },
            ]
          }
        />
      </View>
    </View>
  </View>
</View>
`;
